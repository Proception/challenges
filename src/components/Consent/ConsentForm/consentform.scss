.form-container {
  display: flex;
  justify-content: center;
}

.consent-form {
  
}

.input-container{
  display: flex;
  justify-content: center;
  margin: 15px;
}

.input-field {
  padding: 5px;
  margin-right: 10px;
  border-radius: 5px;
  font-family: Roboto;
  font-style: normal;
  font-weight: normal;
  font-size: 15px;
  color: #212529;
  text-shadow: 0px 0px 0px #595959;
  -webkit-text-fill-color: transparent;
  background: #F7F7F7;
  border: 1px solid #dfe1e5;
  box-sizing: border-box;
}

.input-field:hover {
  box-shadow: 0 1px 6px 0 rgba(32,33,36,0.28);
  border-color: rgba(223,225,229,0);
}

.button-container {
  display: flex;
  justify-content: center;
}

.give-consent-button{
  border-radius: 10px;
  padding: 15px;
  box-shadow: 1px 2px 5px rgba(0, 0, 0, 0.25);
  font-weight: bold;
  background-color: #2867B2;
  color: white;
  transition: box-shadow 0.5s ease-in;
}

.give-consent-button:hover{
  font-weight: bold;
  color: #2867B2;
  background-color: white;
  box-shadow: 0 0px 0px 0 rgba(32,33,36,0.28);
  transition: box-shadow 0.3s ease-in-out;
}

button:disabled {
  color: #2867B2;
  background-color: white;
  cursor: not-allowed;
  box-shadow: 0 0px 0px 0 rgba(32,33,36,0.28);
}

.consent-container{
  display: flex;
  justify-content: center;
  margin: 15px;
  font-family: Roboto;
  font-style: normal;
  font-weight: normal;
  font-size: 15px;
  color: #212529;
}

.consent-header-container {

}

.consent-title{
  display: flex;
  justify-content: center;
  margin: 15px;
  color: #595959;
}

.consent-list{
  padding: 15px;
  border: 1px solid #dfe1e5;
  border-radius: 10px;
  color: #595959;
}