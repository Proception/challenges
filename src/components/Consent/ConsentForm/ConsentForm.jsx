import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import validator from 'validator';
import Input from '../../InputField/Input.jsx';
import Button from '../../Button/Button.jsx';
import Notify from '../../../utils/Notify';
import './consentform.scss';

export default class ConsentForm extends Component {
  constructor(props) {
    super(props);

    this.state = {
      email: '',
      name: '',
      selectedConsents:[]
    };
  }

  generateInputFields = () => {
    const { email, name } = this.state;
    const inputData = [
      {
        name: 'name',
        value: name,
        title: 'Name',
      },
      {
        name: 'email',
        value: email,
        title: 'Email address'
      }
    ];
    return inputData.map(({ name, value, title, type }, index) => {
      return (
        <Fragment key={index}>
          <Input className={'input-field'} name={name} value={value} handleChange={this.handleInput} type={type} placeholder={title} />
        </Fragment>
      );
    });
  }

  generateCheckBoxes = () => {
    const {consents} = this.props;
    
    return consents.map(({ name, value }, index) => {
      return (
        <Fragment key={index}>
          <Input className={'input-field'} name={name} value={value} handleChange={this.handleSelectedConsents} type={'checkbox'}  /> {value} <br/>
        </Fragment>
      );
    });
  } 

  doesEmailExist = (email) => {
    const {consents} = this.props.totalConsents;
    return consents.filter(consent => consent.email === email).length > 0;
  }

  updateConsent = ({name, email, selectedConsents}) => {
    const {viewConsentList} = this.props;
    Notify.update(
      'Would you like to update consent',
      async () => {
        await this.props.updateConsent({name, email, selectedConsents});
        viewConsentList();
      }, 
      ['Cancel', 'Update']
    );
  }

  handleInput = (e) => {
    let value = e.target.value;
    let name = e.target.name;

    this.setState(prevState => {
      return {
        ...prevState, [name]: value
      };
    }, () => { }
    );
  }

  handleSelectedConsents = (e) => {
    const selectedValue = e.target.value;
    const {selectedConsents} = this.state;
    const previouslySelected = selectedConsents.find( consent => selectedValue === consent) !== undefined;

    if (!previouslySelected) {
      this.setState(prevState => {
        return {
          ...prevState, 
          selectedConsents: [...prevState.selectedConsents, selectedValue]
        };
      }, () => {}
      );
    }

    if (previouslySelected) {
      this.setState(prevState => {
        return {
          ...prevState, 
          selectedConsents: [...selectedConsents.filter( consent => selectedValue !== consent)]
        };
      }, () => {}
      );
    }
  }

  handleFormSubmit = async (e) => {
    e.preventDefault();
    const {name, email, selectedConsents} = this.state;
    const {viewConsentList} = this.props;
    const isFormValid = !validator.isEmpty(name) && !validator.isEmpty(email) && validator.isEmail(email) && selectedConsents.length > 0;

    if (!isFormValid) {
      Notify.notifyError('Form is Invalid, please enter a valid email or name');
    }

    if (isFormValid && this.doesEmailExist(email)) {
      this.updateConsent({name, email, selectedConsents});
    }

    if (isFormValid && !this.doesEmailExist(email)) {
      await this.props.saveConsent({name, email, selectedConsents});
      viewConsentList();
    }
  }

  render() {
    return (
      <div className={'form-container'}>
        <form className={'consent-form'} id="consentForm" noValidate onSubmit={this.handleFormSubmit}>
          <div className={'input-container'}>{this.generateInputFields()}</div>
          <div className={'consent-title'}><span>I agree to : </span></div>
          <div className={'consent-container'}>
            <div className={'consent-list'}>{this.generateCheckBoxes()}</div>
          </div>
          <div className={'button-container'}>
            <Button id="consentbutton" title={'Give consent'} classButton={'give-consent-button'} 
              disabled={this.state.selectedConsents.length === 0}
            />
          </div>
        </form>
      </div>
    );
  }
}

ConsentForm.propTypes = {
  rows: PropTypes.array,
  saveConsent: PropTypes.func,
  updateConsent: PropTypes.func,
  viewConsentList: PropTypes.func,
  totalConsents: PropTypes.object,
  consents: PropTypes.array
};
