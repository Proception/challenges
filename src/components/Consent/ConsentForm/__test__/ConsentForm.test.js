import React from 'react';
import sinon from 'sinon';
import renderer from 'react-test-renderer';

import ConsentForm from '../ConsentForm.jsx';
import {
  preventDefault,
  props,
  state,
  consents,
  correctEventSelectedConsents,
  correctEventName,
  correctEventEmail,
  wrongEventEmail,
  wrongEventName
} from '../../../../../__mocks__/consentMocks';

describe('Consent Form Component', () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(<ConsentForm state={state} {...props} />);
  });
  it('renders as expected', () => {
    const component = renderer.create(
      <ConsentForm totalConsents={[]} updateConsent={() => { }} saveConsent={() => { }} consents={[]} />,
    );
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
  it('calls handleInput(e) with the right email and name values', () => {
    sinon.spy(wrapper.instance(), 'handleInput');

    wrapper.instance().handleInput(correctEventEmail);
    wrapper.instance().handleInput(correctEventName);

    expect(wrapper.instance().handleInput.calledTwice)
      .toEqual(true);
    expect(wrapper.instance().handleInput.calledWith(correctEventEmail)).toEqual(true);
    expect(wrapper.instance().handleInput.calledWith(correctEventName)).toEqual(true);
  });
  it('calls handleSelectedConsents(e) with selected values', () => {
    sinon.spy(wrapper.instance(), 'handleSelectedConsents');

    wrapper.instance().handleSelectedConsents(correctEventSelectedConsents);

    expect(wrapper.instance().handleSelectedConsents.calledOnce)
      .toEqual(true);
    expect(
      wrapper.instance().handleSelectedConsents.calledWith(correctEventSelectedConsents)
    ).toEqual(true);
  });
  it('calls handleInput(e) with the wrong email value', () => {
    sinon.spy(wrapper.instance(), 'handleInput');

    wrapper.instance().handleInput(wrongEventEmail);

    expect(wrapper.instance().handleInput.calledOnce)
      .toEqual(true);
    expect(wrapper.instance().handleInput.calledWith(wrongEventEmail)).toEqual(true);
  });
  it('calls handleInput(e) with the wrong name value', () => {
    sinon.spy(wrapper.instance(), 'handleInput');

    wrapper.instance().handleInput(wrongEventName);

    expect(wrapper.instance().handleInput.calledOnce)
      .toEqual(true);
    expect(wrapper.instance().handleInput.calledWith(wrongEventName)).toEqual(true);
  });
  it('should submit successfully with correct input values', () => {
    wrapper = shallow(<ConsentForm {...props} />);
    sinon.spy(wrapper.instance(), 'handleFormSubmit');

    wrapper.instance().handleInput(correctEventEmail);
    wrapper.instance().handleInput(correctEventName);
    wrapper.instance().handleSelectedConsents(correctEventSelectedConsents);

    wrapper.instance().handleFormSubmit(preventDefault);

    expect(wrapper.instance().handleFormSubmit.calledOnce)
      .toEqual(true);
    expect(wrapper.instance().handleFormSubmit.calledWith(preventDefault)).toEqual(true);
  });
  it('should not submit with wrong input values', () => {
    wrapper = shallow(<ConsentForm {...props} />);
    sinon.spy(wrapper.instance(), 'handleFormSubmit');
    wrapper.instance().handleInput(correctEventName);
    wrapper.instance().handleInput(wrongEventEmail);

    wrapper.instance().handleFormSubmit(preventDefault);

    expect(wrapper.instance().handleFormSubmit.calledOnce)
      .toEqual(true);
    expect(wrapper.instance().handleFormSubmit.calledWith(preventDefault)).toEqual(true);
  });
  it('should be able select and deselect consents sucessfully', () => {
    sinon.spy(wrapper.instance(), 'handleSelectedConsents');

    wrapper.instance().handleSelectedConsents(correctEventSelectedConsents);
    wrapper.instance().handleSelectedConsents(correctEventSelectedConsents);

    expect(wrapper.instance().handleSelectedConsents.calledTwice)
      .toEqual(true);
    expect(
      wrapper.instance().handleSelectedConsents.calledWith(correctEventSelectedConsents)
    ).toEqual(true);
  });
});
