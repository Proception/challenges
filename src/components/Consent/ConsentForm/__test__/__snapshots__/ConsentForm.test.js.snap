// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Consent Form Component renders as expected 1`] = `
<div
  className="form-container"
>
  <form
    className="consent-form"
    id="consentForm"
    noValidate={true}
    onSubmit={[Function]}
  >
    <div
      className="input-container"
    >
      <input
        className="input-field"
        name="name"
        onChange={[Function]}
        placeholder="Name"
        value=""
      />
      <input
        className="input-field"
        name="email"
        onChange={[Function]}
        placeholder="Email address"
        value=""
      />
    </div>
    <div
      className="consent-title"
    >
      <span>
        I agree to : 
      </span>
    </div>
    <div
      className="consent-container"
    >
      <div
        className="consent-list"
      />
    </div>
    <div
      className="button-container"
    >
      <button
        className="give-consent-button"
        disabled={true}
        id="consentbutton"
        onClick=""
      >
        <div>
          Give consent
        </div>
      </button>
    </div>
  </form>
</div>
`;
